Index: ctfutils-9.2/sys/cddl/contrib/opensolaris/uts/common/sys/isa_defs.h
===================================================================
--- ctfutils-9.2.orig/sys/cddl/contrib/opensolaris/uts/common/sys/isa_defs.h	2014-10-10 05:32:02.870358340 +0000
+++ ctfutils-9.2/sys/cddl/contrib/opensolaris/uts/common/sys/isa_defs.h	2014-10-10 05:32:19.402358768 +0000
@@ -403,7 +403,7 @@
 #define	_INT_ALIGNMENT			4
 #define	_FLOAT_ALIGNMENT		4
 #define	_FLOAT_COMPLEX_ALIGNMENT	4
-#if defined(__mips_n64)
+#if defined(__mips64) && defined(__LP64__)
 #define	_LONG_ALIGNMENT			8
 #define	_LONG_LONG_ALIGNMENT		8
 #define	_DOUBLE_ALIGNMENT		8
